/**
 * Copyright (c) 2020-present SMC Treviso s.r.l. All rights reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

syntax = "proto3";

import "google/protobuf/struct.proto";

package embedding;

service Embedding {
    rpc GetMessages(EmbeddingRequest) returns (EmbeddingResponse);
}

enum ChunkType {
    CHUNK_TYPE_DEFAULT = 0;
    CHUNK_TYPE_TEXT_SPLITTER = 1;
    CHUNK_TYPE_TOKEN_TEXT_SPLITTER = 2;
    CHUNK_TYPE_CHARACTER_TEXT_SPLITTER = 3;
    CHUNK_TYPE_SEMANTIC_SPLITTER = 4;
}

message RequestChunk {
    optional ChunkType type = 1;
    google.protobuf.Struct jsonConfig = 2;
}

message ProviderModel {
    string provider = 1;
    string model = 2;
}

message EmbeddingModel {
    optional string apiKey = 1;
    ProviderModel providerModel = 2;
    .google.protobuf.Struct jsonConfig = 3;
}

message ResponseChunk {
    int32 number = 1;
    int32 total = 2;
    string text = 3;
    repeated float vectors = 4;
}

message EmbeddingRequest {
    RequestChunk chunk = 1;
    EmbeddingModel embeddingModel = 2;
    string text = 3;
}

message EmbeddingResponse {repeated ResponseChunk chunks = 1;}
